docs/design/kademlia-removal.md

```diff
@@ -0,0 +1,89 @@
+# Kademlia Removal
+
+## Abstract
+
+This design document outlines the communication protocol between satellites and
+storage nodes, network refreshes, and kademlia removal given the satellite opt-in
+capability for storage nodes.
+
+## Background
+
+Many peer-to-peer, decentralized systems employ the Kademlia implementation of a distributed hash table to allow for 
+locating peer nodes, exchanging messages and sharing data. However, due to the nature of our network, we only use Kademlia 
+for node discovery and address lookups given node IDs. This is useful when satellites don’t know about all the nodes in 
+the network and nodes are unfamiliar with all of the satellites in the network.
+
+With our recent business decision of simplifying the storage node operator user experience, we no longer require kademlia 
+for node discovery. In a solution called SNO-select, storage nodes operators manually select the satellites they want to work with 
+and satellites wait for storage nodes to work with them. The initial implementation of this solution allows SNOs to update 
+their trusted satellite list in a hardcoded configuration file, but future improvements will enable users to manage
+this list through a web console. 
+
+We will replace our Kademlia DHT and related entities with direct communication between satellites and storage nodes, 
+and keep the network fresh without kademlia node discovery and random lookups.
+
+## Design
+
+### Nodes reach out to satellites that they want to work with
+- The satellites are listed in the trust package
+- Nodes should communicate with satellites directly rather than using kademlia to traverse the network to find the address of a given ID.
+- Storage nodes should notify satellites when they start up, wait a random amount of time (to add jitter 
+http://highscalability.com/blog/2012/4/17/youtube-strategy-adding-jitter-isnt-a-bug.html), then start reporting in roughly on the hour
+
+### Network refreshes at a regular interval
+- Nodes will keep themselves up to date in the network by pinging all the satellites in their
+   trusted list every hour.
+- Satellites will ping the nodes back to confirm their addresses
+    - If is it successful, the satellite will insert or update the node in the overlay cache and
```

Comment:
(I know this is already merged, and I'm late! But) Might want to note some requirements for a 'successful' pingback:

If a node with nodeID `N` has contacted satellite `S` and claimed its address is `A`,

1. `S` _must_ initiate a network connection `C` to address `A`
2. `S` _must_ verify that the remote endpoint on `C` has a private key corresponding to public key/node ID `N`. (It would be sufficient to complete negotiation of an SSL session over `C` and then verify that the remote end is using the same certificate used by `N` in the initial incoming ping.)
3. `S` _should_ verify that the remote endpoint on `C` agrees that its address is `A`. (This doesn't seem strictly necessary for security, but could prevent misconfigurations where a SNO runs multiple nodes with the same identity.)
4. `S` _should_ respond to the initial incoming ping from `N` with the result of the pingback, so that the dashboard on that node can report whether the node can receive incoming connections.

Reply:
I added that information in https://github.com/storj/storj/pull/2837